ethcore/src/spec/spec.rs

```diff
@@ -540,7 +515,20 @@ impl Spec {
 
 		// TODO: could optimize so we don't re-run, but `ensure_db_good` is barely ever
 		// called anyway.
-		let db = self.run_constructors(factories, db)?;
+		let (root, db) = run_constructors(
+			&self.genesis_state,
+			&self.constructors,
+			&*self.engine,
+			self.author,
+			self.timestamp,
+			self.difficulty,
+			factories,
+			db
+		)?;
+		if root != self.state_root() {
+			return Err("Invalid state root".into());
```

Comment:
this should actually never occur, unless someone modifies this code and at some point overwrites `state_root_memo`. Maybe I should change it to `assert_eq!`?